### ISSUE TITLE
`#cite` and `#bibliography` do not work inside a `#locate` call

### ISSUE THREAD

**Issue by andreasKroepelin** [typst/typst#785]
Consider having a file `literature.bib`
```bib
@misc{key,
    author = "Author",
    year = "2023"
}
```

Then this works:
```typ
#cite("key")
#bibliography("literature.bib")
```
and this doesn't work:
```typ
#locate(loc => {
    cite("key")
})
#locate(loc => {
    bibliography("literature.bib")
})
```

The error is `the document does not contain a bibliography`.

**Comment by TeddyHuang-00** [comment]
Just want to add that reference to labels also won't work in such situation. I guess this is related and may can be fixed altogether.

**Comment by LarsKue** [comment]
Possibly related to #622.

**Comment by leo848** [comment]
Is there any known workaround, especially regarding citations in `locate` calls (#785)? Since locate is necessary for `state`, this would be pretty essential for what I'm trying to achieve.

**Comment by laurmaedje** [comment]
> Is there any known workaround

None that I know of at the moment, sadly

**Comment by Dherse** [comment]
As self-appointed necromancer of GitHub issues, there is a way of solving this: one can add a `potential_errors` in the introspection loop (`typset` in `src/model/mods.rs`) outside of the loop, every iteration it gets reset `potential_errors.clear()` and then you can add it as a parameter to the `Show` trait. When you exit the loop, you can check whether the list of errors is empty, if it's not then you failed compilation.

I think that this has several advantages:
1. More than one error can be shown to the user
2. it "delays" error until introspection is done which fixes a class of bugs relating to introspection
3. I am god and I bless this solution therefore it is good 😂

**Comment by andreasKroepelin** [comment]
@Dherse This allows you to show multiple errors, fine. And I think there are multiple ones here. But how does that help with having `cite` or `bibliography` inside a `locate`?

**Comment by Dherse** [comment]
@andreasKroepelin The reason why this happens, is because when you're in a `locate` block, you need introspection to be ready to be able to create the `Content` of that `locate` block. Because of this, when citing, it can try and look for a `bibliography` too early in the introspection loop (that loop in `src/model/mods.rs` runs as many times as needed for layouting to be done, with a maximum of five loops).

What my fix allows is that if in an iteration of the loop we can't find the `label` or citation to reference, we will generate an error in `potential_errors`, but we will only "throw" those errors once we exit the loop. What this does is allow `Show` in `CiteElem` to throw an error without halting the loop, letting it run **until** introspection is:
- either fully done, in which case the citation doesn't exist and we throw the error for real
- introspection is done "enough" that we **can** find the citation and display it

This is how it fixes the bug, and I have tested, it works as expected: (element here means citation or label)
- if the element exist, it no longer throws an error, instead waits for introspection to be "ready"
- if the element doesn't exist, it throws an error once introspection is done

It literally delays this specific error until introspection is done and we are "sure" that it is indeed an error and not just a lack of information, which was the cause of the bug previously.

**Comment by andreasKroepelin** [comment]
Alright, that makes sense. Thanks a lot for the clarification!

**Comment by ModProg** [comment]
> and I have tested, it works as expected
@Dherse do you have your test in some branch I could try? Would really like to be able to have bibliography in typs-slides and would be fine running a hacked together solution for now.

**Comment by laurmaedje** [comment]
I agree that this is quite important to fix. I will look into it, so that hopefully we can have it fixed for version 0.6.

**Comment by laurmaedje** [comment]
I implemented the approach outlined by @Dherse. The problem should be fixed on main and will ship in 0.6.
